packages/vm/src/testing/storage.rs

```diff
@@ -15,44 +19,45 @@ const GAS_COST_LAST_ITERATION: u64 = 37;
 #[cfg(feature = "iterator")]
 const GAS_COST_RANGE: u64 = 11;
 
-/// A storage iterator for testing only. This type uses a Rust iterator
-/// as a data source, which does not provide a gas value for the last iteration.
-#[cfg(feature = "iterator")]
-pub struct MockIterator<'a> {
-    source: Box<dyn Iterator<Item = (KV, u64)> + 'a>,
-}
-
 #[cfg(feature = "iterator")]
-impl MockIterator<'_> {
-    pub fn empty() -> Self {
-        MockIterator {
-            source: Box::new(std::iter::empty()),
-        }
-    }
-}
-
-#[cfg(feature = "iterator")]
-impl StorageIterator for MockIterator<'_> {
-    fn next(&mut self) -> FfiResult<Option<KV>> {
-        match self.source.next() {
-            Some((kv, gas_used)) => (Ok(Some(kv)), GasInfo::with_externally_used(gas_used)),
-            None => (
-                Ok(None),
-                GasInfo::with_externally_used(GAS_COST_LAST_ITERATION),
-            ),
-        }
-    }
+#[derive(Default, Debug)]
+struct Iter {
+    data: Vec<KV>,
+    position: usize,
 }
 
 #[derive(Default, Debug)]
 pub struct MockStorage {
     data: BTreeMap<Vec<u8>, Vec<u8>>,
+    #[cfg(feature = "iterator")]
+    iterators: HashMap<u32, Iter>,
 }
 
 impl MockStorage {
     pub fn new() -> Self {
         MockStorage::default()
     }
+
+    #[cfg(feature = "iterator")]
+    pub fn all(&mut self, iterator_id: u32) -> FfiResult<Vec<KV>> {
```

Comment:
Ah, this is just on MockStorage not on Storage.
But we get to use it in all the tests cuz of generics not loosing info on the specific types.

Nice.